kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,617 @@
+
+    using TGPMapCommunicator = GlobalPointerMapCommunicator<TPointerDataType, TValueDataType>;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    /**
+     * @brief Construct a new Apply Proxy object
+     *
+     * @param CurrentRank               Current rank
+     * @param rLocalApplyFunctor        Thread safe update lambda method
+     * @param rNonLocalApplyFunctor     Thread safe non-local gp map value update method
+     * @param rNonLocalDataMap          Non-local gp map
+     * @param rPointerCommunicator      Map communicator
+     */
+    ApplyProxy(
+        const int CurrentRank,
+        const TLocalApplyFunctor& rLocalApplyFunctor,
+        const TNonLocalApplyFunctor& rNonLocalApplyFunctor,
+        TGPNonLocalDataMap& rNonLocalDataMap,
+        TGPMapCommunicator& rPointerCommunicator)
+        : mCurrentRank(CurrentRank),
+          mrLocalApplyFunctor(rLocalApplyFunctor),
+          mrNonLocalApplyFunctor(rNonLocalApplyFunctor),
+          mrNonLocalDataMap(rNonLocalDataMap),
+          mrPointerCommunicator(rPointerCommunicator)
+    {
+        // identify proper methods to be used in updates
+        if (mrPointerCommunicator.IsDistributed()) {
+            this->mUpdateMethod = &ProxyType::UpdateLocalAndRemoteData;
+        } else {
+            this->mUpdateMethod = &ProxyType::UpdateLocalData;
+        }
+    }
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    /**
+     * @brief Updates values with given values map
+     *
+     * This method updates local gps using mrLocalApplyFunctor instantly. Value updates
+     * belonging to remote gps are stored in the mrNonLocalDataMap using mrNonLocalApplyFunctor.
+     *
+     * In the case of serial job, this method calls UpdateLocalData, where no checks are done
+     * to ensure gps are local because all gps are local.
+     *
+     * In case of distributed job, this method calls UpdateLocalAndRemoteData where checks are
+     * performed to identify gps are local or non-local, based on that mrLocalApplyFunctor or mrNonLocalApplyFunctor
+     * methods are called.
+     *
+     * The input rGPDataMap map should not contain any key gps which are not included in creating the
+     * GlobalPointerMapCommunicator used in this proxy (i.e. mrPointerCommunicator)
+     *
+     * This method can be called several times without severe additional computational cost
+     * since this does not do any mpi communication.
+     *
+     * This does not have additional cost in serial run
+     *
+     * @see GlobalPointerMapCommunicator
+     *
+     * @param rGPDataMap        Input values map (key: GlobalPointer<TPointerDataType>, value: TValueDataType)
+     */
+    void Update(const TGPDataMap& rGPDataMap)
+    {
+        // get gp vector
+        const auto& gps = TGPMapCommunicator::GetKeys(rGPDataMap);
+
+        // running this in parallel assuming mrLocalApplyFunctor is thread safe
+        IndexPartition<int>(gps.size()).for_each([&](const int Index) {
```

Comment:
what happens if the GPDataMap does not touch all the entries in the gp_list? what if it touches some that still did not exist?

I think you should either replace all of the map at onces, or only allow to change the values (note that for efficiency of repeated calls you need to compute the communication pattern before the apply function)

Reply:
If GPDataMap does not touch the entire gp_list, then it make sense to compute the communication plan in `SendAndApplyRemotely` method because that is where mpi communication is happening. If we are to do this,

Option 1: As you suggest replace all of the map at once. Then create the communication plan, and then apply (both of these should be done in `SendAndApplyRemotely`. (Not a big fan of this because, I have situations where I have to call Update method few different times, and needs to keep values after each update method. So if this is to be done, then each `Update` method needs to be followed by `SendAndApplyRemotely` call which is expensive but less fragile)

Option 2: Update an existing map, Then create the communication plan, and then apply (both of these should be done in `SendAndApplyRemotely`. This is going to be little bit expensive if not used wisely (explained later).

Option 3: Get rid of the map update, only allow update via Gps. So this will create a non_local_gp_map inside the communicator. After all the update calls, then create the communication plan, and then apply (both of these should be done in `SendAndApplyRemotely`. (I like this one better, because this gives more flexibility to user, and user does not need to maintain a gp_map)

> If GPDataMap touches some gps which still does not exist

I am not sure how this can happen. Because GPs are made out of actual nodes/elements/conditions, so the actual entity should exist in some rank. In the current implementation. If you meant that GPDataMap touches an gp which is not in the non_local_gp_map, then following will happen in each option

Option 1: It won't be a problem, since the total non_local_gp_map is always regenerated from the given map.

Option 2: This may need to extend non_local_gp_map as well, which is expensive. But won't give any errors since communication plan and communication is done after all `Update` calls. (It won't be expensive at all, if user provides all the gps which they may require for communication when creating the `GlobalPointerMapCommunicator`. So it is users discretion to be efficient or not)

Option 3: This may also need to extend non_local_gp_map. But won't give any errors since communication plan and communication is done after all `Update` calls. (It won't be expensive at all, if user provides all the gps which they may require for communication when creating the `GlobalPointerMapCommunicator`. So it is users discretion to be efficient or not)

Overall I like allowing Option 2 and/or 3. Option 2 and 3 because, user can make it efficient (If user uses it wisely, then it will be the most efficient in memory consumption as well as execution time). Option 3 is better than option 2 because, if we only allow option 3, then user does not need keep a gp map. 

But if he user does not wisely get the advantage of passing all the gps to `GlobalPointerMapCommunicator` at constructor, then it is going to be little bit expensive. I am not in favor of Option 1 mere because, I will need to call `Update` method more than once, then I need to avoid doing expensive mpi communication everytime while keeping the track of the values passed through `Update`.

I am not sure I uderstood your question perfectly. But hope this clarifies some, and would be great to hear your opinion. :)

Reply:
To explain my concern let's consider the following scenario

CPU0 has nodes 1 2 3
CPU1 has nodes 4 5 6
CPU2 has nodes 7 8 9

let's imagine that CPU0 wants to Add a value to the nodes 1,5,9 while the others do nothing.
this is a completely admissible scenario. In that case CPU0 would gather the pointer gp1,gp5,gp9 into gp_list while gp_list would be empty for all the others.

My concern is that someone may try to set the map
{{1,1.0}{5,5.0}{9,5.0},{7,5.0}}

where the entry 7 was originally in the list. This should be disallowed (easily accomplished if you use find() instead of [] in the setting up operator}. Also you should loop over the cp_list you memorize instead of on the map you passed. That way if something is not in the map it will be detected.

i am ok about the function which sets the map one by one.

Reply:
done as discussed :)